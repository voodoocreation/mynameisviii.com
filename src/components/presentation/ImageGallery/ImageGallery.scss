@import "../../../scss/variables", "../../../scss/mixins";

.ImageGallery {
  position: relative;

  figcaption {
    @include hide-visually();
  }

  .Image {
    @include square-image();

    opacity: 0.7;
    transition: opacity 200ms ease-out;

    &:hover {
      opacity: 1;
    }
  }

  &--modal {
    img {
      position: absolute;
      top: 50%;
      left: 50%;
      width: 100%;
      height: 100%;
      transform: translate3d(-50%, -50%, 0);
      object-fit: scale-down;
    }

    figcaption {
      @include panel-1();

      position: absolute;
      right: 0;
      bottom: 0;
      left: 0;
      text-align: center;
      padding: $spacing;
      opacity: 0.5;
      transition: opacity 200ms ease-out;

      &:hover {
        opacity: 1;
      }
    }

    button {
      &:focus,
      &:active {
        color: $link-color-hover;
      }
    }

    .Loader {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate3d(-50%, -50%, 0);
    }

    &--previousButton,
    &--nextButton {
      @include text-shadow();

      position: absolute;
      top: 50%;
      font-size: 28px;
      padding: $spacing;
      transform: translate3d(0, -50%, 0);
    }

    &--previousButton {
      left: 0;
    }

    &--nextButton {
      right: 0;
    }
  }
}
