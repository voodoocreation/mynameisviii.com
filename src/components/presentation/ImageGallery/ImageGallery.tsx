import cn from "classnames";
import * as React from "react";
import { MdArrowBack, MdArrowForward } from "react-icons/md";

import Modal from "../Modal/Modal";

interface IProps {
  className?: string;
  isLooped?: boolean;
  usePortal?: boolean;
  onItemClick?: (index: number) => void;
  onGoTo?: (index: number) => void;
  onModalClose?: () => void;
  onNext?: (index: number) => void;
  onPrevious?: (index: number) => void;
}

interface IState {
  currentIndex?: number;
  isOpen: boolean;
}

export default class ImageGallery extends React.Component<IProps, IState> {
  public static defaultProps = {
    isLooped: true,
    usePortal: true
  };

  public readonly state = {
    currentIndex: undefined,
    isOpen: false
  };

  public render() {
    const { children, className } = this.props;
    const { currentIndex } = this.state;

    const currentItem =
      currentIndex !== undefined ? this.getItemAtIndex(currentIndex) : null;

    return (
      <React.Fragment>
        <div className={cn("ImageGallery", className)}>
          {React.Children.map(children, (item, index) =>
            React.cloneElement(item as React.ReactElement<any>, {
              onClick: this.onItemClick(index)
            })
          )}
        </div>

        <Modal
          className="ImageGallery-modal"
          isOpen={this.state.isOpen}
          usePortal={this.props.usePortal}
          onClose={this.onModalClose}
          onKeyDown={this.onModalKeyDown}
        >
          <div className="ImageGallery-modal-item">{currentItem}</div>

          <div className="ImageGallery-modal-controls">
            <button
              className="ImageGallery-modal-previous"
              onClick={this.previous}
            >
              <MdArrowBack />
            </button>

            <button className="ImageGallery-modal-next" onClick={this.next}>
              <MdArrowForward />
            </button>
          </div>
        </Modal>
      </React.Fragment>
    );
  }

  public goTo = (nextItemIndex: number) => {
    if (this.getItemAtIndex(nextItemIndex)) {
      this.setState({
        currentIndex: nextItemIndex
      });
    }

    if (this.props.onGoTo) {
      this.props.onGoTo(nextItemIndex);
    }
  };

  public previous = () => {
    const { currentIndex } = this.state;
    const previousIndex = currentIndex !== undefined ? currentIndex - 1 : 0;
    const goToIndex =
      this.props.isLooped && !this.getItemAtIndex(previousIndex)
        ? React.Children.count(this.props.children) - 1
        : previousIndex;

    this.goTo(goToIndex);

    if (this.props.onPrevious) {
      this.props.onPrevious(goToIndex);
    }
  };

  public next = () => {
    const { currentIndex } = this.state;
    const nextIndex = currentIndex !== undefined ? currentIndex + 1 : 0;
    const goToIndex =
      this.props.isLooped && !this.getItemAtIndex(nextIndex) ? 0 : nextIndex;

    this.goTo(goToIndex);

    if (this.props.onNext) {
      this.props.onNext(goToIndex);
    }
  };

  private getItemAtIndex = (findIndex: number) =>
    React.Children.toArray(this.props.children).find(
      (_, index) => index === findIndex
    );

  private onItemClick = (index: number) => () => {
    this.setState({
      currentIndex: index,
      isOpen: true
    });

    if (this.props.onItemClick) {
      this.props.onItemClick(index);
    }
  };

  private onModalKeyDown = (event: KeyboardEvent) => {
    switch (event.key) {
      case "ArrowLeft":
        this.previous();
        break;

      case "ArrowRight":
        this.next();
        break;
    }
  };

  private onModalClose = () => {
    this.setState({
      currentIndex: undefined,
      isOpen: false
    });

    if (this.props.onModalClose) {
      this.props.onModalClose();
    }
  };
}
