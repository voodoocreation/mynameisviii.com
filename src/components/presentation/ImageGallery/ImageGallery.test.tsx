import { mount, render } from "enzyme";
import * as React from "react";

import ImageGallery from "./ImageGallery";

const setup = (fn: any, fromTestProps?: any) => {
  const props = {
    className: "TestGallery",
    usePortal: false,
    ...fromTestProps
  };

  return {
    actual: fn(
      <ImageGallery {...props}>
        <div className="Image-1" />
        <div className="Image-2" />
        <div className="Image-3" />
      </ImageGallery>
    ),
    props
  };
};

describe("[presentation] <ImageGallery />", () => {
  it("renders correctly with minimum props", () => {
    const { actual } = setup(render);

    expect(actual).toMatchSnapshot();
  });

  it("opens modal and shows correct item when an item is clicked", () => {
    const { actual } = setup(mount);

    actual.find(".Image-2").simulate("click");
    expect(actual.state("isOpen")).toBe(true);
    expect(actual.state("currentIndex")).toBe(1);
  });

  it("calls `onModalClose` prop when defined and the modal is closed", () => {
    const { actual, props } = setup(mount, { onModalClose: jest.fn() });

    actual.find(".Image-2").simulate("click");
    actual.find(".Modal-closeButton").simulate("click");
    expect(props.onModalClose).toHaveBeenCalled();
  });

  it("doesn't throw an error when `onModalClose` prop is undefined and the modal is closed", () => {
    let isPassing = true;
    const { actual } = setup(mount);

    try {
      actual.find(".Image-2").simulate("click");
      actual.find(".Modal-closeButton").simulate("click");
    } catch (error) {
      isPassing = false;
    }

    expect(isPassing).toBe(true);
  });

  describe("keyboard events", () => {
    it("left arrow triggers previous() correctly", () => {
      const { actual } = setup(mount);

      actual.find(".Image-2").simulate("click");
      window.dispatchEvent(new KeyboardEvent("keydown", { key: "ArrowLeft" }));
      expect(actual.state("currentIndex")).toBe(0);
    });

    it("right arrow triggers next() correctly", () => {
      const { actual } = setup(mount);

      actual.find(".Image-1").simulate("click");
      window.dispatchEvent(new KeyboardEvent("keydown", { key: "ArrowRight" }));
      expect(actual.state("currentIndex")).toBe(1);
    });
  });

  describe("goTo()", () => {
    it("updates `currentIndex` correctly", () => {
      const { actual } = setup(mount);

      actual.instance().goTo(1);
      expect(actual.state("currentIndex")).toBe(1);
    });

    it("doesn't change `currentIndex` when requested item doesn't exist", () => {
      const { actual } = setup(mount);

      actual.instance().goTo(3);
      expect(actual.state("currentIndex")).toBeUndefined();
    });

    it("calls `onGoTo` prop when defined", () => {
      const { actual, props } = setup(mount, { onGoTo: jest.fn() });

      actual.instance().goTo(2);
      expect(actual.state("currentIndex")).toBe(2);
      expect(props.onGoTo).toHaveBeenCalledWith(2);
    });
  });

  describe("previous()", () => {
    it("updates `currentIndex` correctly", () => {
      const { actual } = setup(mount);

      actual.instance().goTo(2);
      actual.instance().previous();
      expect(actual.state("currentIndex")).toBe(1);
    });

    it("goes to first item when `currentIndex` is undefined", () => {
      const { actual } = setup(mount);

      actual.instance().previous();
      expect(actual.state("currentIndex")).toBe(0);
    });

    it("goes to last item when at the first item and isLooped=true", () => {
      const { actual } = setup(mount);

      actual.instance().goTo(0);
      actual.instance().previous();
      expect(actual.state("currentIndex")).toBe(2);
    });

    it("doesn't change `currentIndex` when at the first item and isLooped=false", () => {
      const { actual } = setup(mount, { isLooped: false });

      actual.instance().goTo(0);
      actual.instance().previous();
      expect(actual.state("currentIndex")).toBe(0);
    });

    it("calls `onPrevious` prop when defined", () => {
      const { actual, props } = setup(mount, { onPrevious: jest.fn() });

      actual.instance().goTo(1);
      actual.instance().previous();
      expect(props.onPrevious).toHaveBeenCalledWith(0);
    });
  });

  describe("next()", () => {
    it("updates `currentIndex` correctly", () => {
      const { actual } = setup(mount);

      actual.instance().goTo(0);
      actual.instance().next();
      expect(actual.state("currentIndex")).toBe(1);
    });

    it("goes to first item when `currentIndex` is undefined", () => {
      const { actual } = setup(mount);

      actual.instance().next();
      expect(actual.state("currentIndex")).toBe(0);
    });

    it("goes to first item when at the last item and isLooped=true", () => {
      const { actual } = setup(mount);

      actual.instance().goTo(2);
      actual.instance().next();
      expect(actual.state("currentIndex")).toBe(0);
    });

    it("doesn't change `currentIndex` when at the last item and isLooped=false", () => {
      const { actual } = setup(mount, { isLooped: false });

      actual.instance().goTo(2);
      actual.instance().next();
      expect(actual.state("currentIndex")).toBe(2);
    });

    it("calls `onNext` prop when defined", () => {
      const { actual, props } = setup(mount, { onNext: jest.fn() });

      actual.instance().goTo(1);
      actual.instance().next();
      expect(props.onNext).toHaveBeenCalledWith(2);
    });
  });
});
