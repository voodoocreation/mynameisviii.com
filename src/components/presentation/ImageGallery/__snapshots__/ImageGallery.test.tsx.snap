// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <ImageGallery /> renders correctly with minimum props 1`] = `
<div
  class="ImageGallery TestGallery"
>
  <div
    class="Image-1"
  />
  <div
    class="Image-2"
  />
  <div
    class="Image-3"
  />
</div>
`;
